embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/comms/delivery/EmbraceDeliveryService.kt

```diff
@@ -191,8 +194,18 @@ internal class EmbraceDeliveryService(
         }
     }
 
-    private fun getFailedSpanEndTimeMs(envelope: Envelope<SessionPayload>) =
-        envelope.getSessionSpan()?.endTimeNanos?.nanosToMillis() ?: -1
+    /**
+     * To approximate the time of any snapshot to be converted into a failed span, we look to the session span of the payload and take
+     * either the end time or the last heartbeat time, whichever exists and is later. If the session span itself is a snapshot, it will
+     * not have an end time, in which case it will fall back to the last heartbeat time. If either exists, it means we can't find a better
+     * time, so we just leave it at 0.
+     */
+    private fun getFailedSpanEndTimeMs(envelope: Envelope<SessionPayload>): Long {
```

Comment:
I found this a bit hard to follow, would suggest breaking it down with local variables. E.g.

```
    private fun getFailedSpanEndTimeMs(envelope: Envelope<SessionPayload>): Long {
        val sessionSpan = envelope.getSessionSpan()
        val endTime = sessionSpan?.endTimeNanos ?: 0L
        val heartbeatTime = sessionSpan?.attributes?.findAttributeValue(embHeartbeatTimeUnixNano.attributeKey.key)?.toLongOrNull() ?: 0L
        return max(endTime, heartbeatTime).nanosToMillis()
    }
```

Reply:
yeah good idea. I'll do that!

I also want to pull this out of this class because it's not really a delivery concern, but I didn't want to do that now, or when I rejiggered the logic initially. I should do it soon....